src/python/gem5/simulate/simulator.py

```diff
@@ -353,10 +353,22 @@ def print_hello() -> bool:
                     # In instances where we have a list of functions, we
                     # convert this to a generator.
                     self._on_exit_event[key] = (func() for func in value)
-                elif isinstance(value, Callable):
+                elif isinstance(value, function):
                     # In instances where the user passes a lone function, the
                     # function is called on every exit event of that type. Here
                     # we convert the function into an infinite generator.
+
+                    # We check if the function is a generator. If it is we
+                    # throw a warning as this is likely a mistake.
+                    if isinstance(value(), Generator):
```

Comment:
This will break when you pass in a function since it calls the function.

You should be able to use https://docs.python.org/3/library/inspect.html#inspect.isgeneratorfunction to check if the thing is a pointer to a function that is a generator.

So, the code should be something like

```python
import inspect
if inspect.isgeneratorfunction(value):
```

Reply:
Impressively, our tests caught this! Good job writing tests!!!

https://github.com/gem5/gem5/actions/runs/10508905940/job/29134080005?pr=1499#step:7:89

Reply:
`Test: simulator-exit-event-handler-with-lone-function-ALL-x86_64-opt Failed`

Wow. Nice and precise too. I'm quite proud of that.

Reply:
> This will break when you pass in a function since it calls the function.
> 
> You should be able to use https://docs.python.org/3/library/inspect.html#inspect.isgeneratorfunction to check if the thing is a pointer to a function that is a generator.
> 
> So, the code should be something like
> 
> ```python
> import inspect
> if inspect.isgeneratorfunction(value):
> ```

This worked perfectly. i've incorporated this into the PR.